src/python/pants/goal/completion.py

```diff
@@ -0,0 +1,184 @@
+
+        The trigger to determine which case we're in is the presence of the passthrough arguments. If there are passthrough
+        arguments, then we're generating completion options. If there are no passthrough arguments, then we're generating
+        a completion script.
+        """
+        if options._passthru:
+            completion_options = self._generate_completion_options(options)
+            if completion_options:
+                print("\n".join(completion_options))
+            return PANTS_SUCCEEDED_EXIT_CODE
+
+        script = self._generate_completion_script(self.shell)
+        print(script)
+        return PANTS_SUCCEEDED_EXIT_CODE
+
+    def _generate_completion_script(self, shell: Shell) -> str:
+        """Generate a completion script for the specified shell.
+
+        Implementation note: In practice, we're just going to read in
+        and return the contents of the appropriate static completion script file.
+
+        :param shell: The shell to generate a completion script for.
+        :return: The completion script for the specified shell.
+        """
+        if shell == Shell.ZSH:
+            return read_resource(_COMPLETIONS_PACKAGE, "pants-completion.zsh").decode("utf-8")
+        else:
+            return read_resource(_COMPLETIONS_PACKAGE, "pants-completion.bash").decode("utf-8")
+
+    def _generate_completion_options(self, options: Options) -> list[str]:
+        """Generate the completion options for the specified args.
+
+        We're guaranteed to have at least two arguments (`["pants", ""]`). If there are only two arguments,
+        then we're at the top-level Pants command, and we should show all goals.
+        - `pants <tab>` -> `... fmt fix lint list repl run test ...`
+
+        If we're at the top-level and the user has typed a hyphen, then we should show global options.
+        - `pants -<tab>` -> `... --pants-config-files --pants-distdir --pants-ignore ...`
+
+        As we add goals, we should show the remaining goals that are available.
+        - `pants fmt fix lint <tab>` -> `... list repl run test ...`
+
+        If there is a goal in the list of arguments and the user has typed a hyphen, then we should
+        show the available scoped options for the previous goal.
+        - `pants fmt -<tab>` -> `... --only ...`
+
+        :param options: The options object for the current Pants run.
+        :return: A list of completion options.
+        """
+        logger.debug(f"Completion passthrough options: {options._passthru}")
+        args = [arg for arg in options._passthru if arg != "pants"]
+        current_word = args.pop()
+        previous_goal = self._get_previous_goal(args)
+        logger.debug(f"Current word is '{current_word}', and previous goal is '{previous_goal}'")
+
+        all_goals = sorted([k for k, v in options.known_scope_to_info.items() if v.is_goal])
```

Comment:
I think we could refactor the current word startswith `-` checks to be before we gather this `all_goals` as it's not needed in those cases.

Reply:
Yep, you're right - the location was legacy, from an earlier iteration

Reply:
I'm going to punt this to https://github.com/pantsbuild/pants/issues/20676 as well, due to the fact that I'll already be trying to remove swathes of code in there with some of the Options built-ins re:

```python
# TODO: This was written as a port of the original bash script, but since we have
# more knowledge of the options and goals, we can make this more robust and accurate (after tests are written).
```

Might end up changing the structure entirely.